syntax = "proto2";

import "serialization/numerics.proto";
import "serialization/quantities.proto";

package principia.serialization;

message Integrator {
  extensions 3000 to 3999;  // Last used: 3001.
}

message AdaptiveStepSizeIntegrator {
  extend Integrator {
    optional AdaptiveStepSizeIntegrator extension = 3000;
  }
  enum Kind {
    DORMAND_ELMIKKAWY_PRINCE_1986_RKN_434FM = 1;
  }
  required Kind kind = 1;
}

message FixedStepSizeIntegrator {
  extend Integrator {
    optional FixedStepSizeIntegrator extension = 3001;
  }
  enum Kind {
    DUMMY = 0;  // For testing.
    BLANES_MOAN_2002_SRKN_6B = 1;
    BLANES_MOAN_2002_SRKN_11B = 2;
    BLANES_MOAN_2002_SRKN_14A = 3;
    MCLACHLAN_1995_SB3A_4 = 4;
    MCLACHLAN_1995_SB3A_5 = 5;
    MCLACHLAN_ATELA_1992_ORDER_4_OPTIMAL = 6;
    MCLACHLAN_ATELA_1992_ORDER_5_OPTIMAL = 7;
    OKUNBOR_SKEEL_1994_ORDER_6_METHOD_13 = 8;
    QUINLAN_1999_ORDER_8A = 9;
    QUINLAN_1999_ORDER_8B = 10;
    QUINLAN_TREMAINE_1990_ORDER_8 = 11;
    QUINLAN_TREMAINE_1990_ORDER_10 = 12;
    QUINLAN_TREMAINE_1990_ORDER_12 = 13;
    QUINLAN_TREMAINE_1990_ORDER_14 = 14;
  }
  required Kind kind = 1;
}

message IntegratorInstance {
  extensions 7000 to 7999;  // Last used: 7001

  required SystemState current_state = 1;
}

//TODO(phl):Adaptive.

message SymmetricLinearMultistepIntegratorInstance {
  extend IntegratorInstance {
    optional SymmetricLinearMultistepIntegratorInstance extension = 7000;
  }
  message Step {
    repeated DoublePrecision displacements = 1;
    repeated DoublePrecision accelerations = 2;
    required DoublePrecision time = 3;
  }
  repeated Step previous_steps = 1;
  required Quantity step = 2;
}

message SymplecticRungeKuttaNystromIntegratorInstance {
  extend IntegratorInstance {
    optional SymplecticRungeKuttaNystromIntegratorInstance extensions = 7001;
  }
  required Quantity step = 1;
}

message SystemState {
  repeated DoublePrecision position = 1;
  repeated DoublePrecision velocity = 2;
  required DoublePrecision time = 3;
}
